### ISSUE TITLE
Unable to search in knowledge base

### ISSUE THREAD

**Issue by paidge** [pluginsGLPI/formcreator#2634]
**Describe the bug**
In the simplified service catalog (self-service role), if knowledge base and forms are splitted, a search input is displayed for the knowledge base but it does not work. A message indicates: "No form yet in this category"

**To Reproduce**
1. Split knowledge base and forms (Administration > Entities > My Entity > Forms > Distinct menu entry for knowledge base)
2. Go to the simplified service catalog (role = self-service) to create a ticket or to search in the knowledge base
3. Click on the knowledge base and type in the search input to filter articles
4. See the message indicating: "No form yet in this category"

**Expected behavior**
This search input should filter articles in the FAQ and not the forms.

**Screenshots**
![image](https://user-images.githubusercontent.com/2470104/157426479-7dfa81f6-3650-4c8c-be6a-4f04eb54d0f8.png)

**GLPI / Plugins (please complete the following information):**
- GLPI version 9.5.7
- list of all plugins : formCreator, genericobject

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Firefox 97

**Comment by paidge** [comment]
Yep ;)

**Comment by btry** [comment]
Hi

Which version of Formcreator are you using ?

**Comment by paidge** [comment]
v2.12.5

**Comment by btry** [comment]
Hi

I did a test, it seems to be working for me. 

Can you check if you have an error in the logs ? Logs are located in glpi/files/_log/. Files to check are php-errors.log and sql-errors.log.

Also, open the developer tools of your browser, then try to do a search for your FAQ items. Check in the networking tab if you have HTTP requests with code >= 300.

**Comment by paidge** [comment]
Hi. Here is the only error I can find in sql-errors.log : 
```[2022-03-07 16:43:16] glpisqllog.ERROR: DBmysql::query() in /var/www/apps/glpi/inc/dbmysql.class.php line 335
  *** MySQL query error:
  SQL: SELECT `glpi_plugin_formcreator_forms`.`id`, `glpi_plugin_formcreator_forms`.`name`, `glpi_plugin_formcreator_forms`.`icon`, `glpi_plugin_formcreator_forms`.`icon_color`, `glpi_plugin_formcreator_forms`.`background_color`, `glpi_plugin_formcreator_forms`.`descrip>
                  AGAINST('comment* on* s\\\'en* sert* ?* ;)*' IN BOOLEAN MODE) OR MATCH(glpi_plugin_formcreator_questions.`name`, glpi_plugin_formcreator_questions.`description`)
                  AGAINST('comment* on* s\\\'en* sert* ?* ;)*' IN BOOLEAN MODE))) AND ((`access_rights` != '2' OR (`glpi_plugin_formcreator_forms_profiles`.`profiles_id` = '1')))) GROUP BY `glpi_plugin_formcreator_forms`.`id`, `glpi_plugin_formcreator_forms`.`name`, `gl>
  Error: syntax error, unexpected ')', expecting $end
  Backtrace :
  inc/dbmysqliterator.class.php:95
  inc/dbmysql.class.php:902                          DBmysqlIterator->execute()
  plugins/formcreator/inc/form.class.php:983         DBmysql->request()
  plugins/formcreator/ajax/homepage_wizard.php:62    PluginFormcreatorForm->showFormList()
  plugins/formcreator/ajax/homepage_wizard.php:48    plugin_formcreator_showWizardForms()
```

If I click on See all, I can't see any FAQ item : 
![image](https://user-images.githubusercontent.com/2470104/159227409-cd643ed2-eb23-47b0-8f1a-b26343841fba.png)

If I click on the category, I can see my FAQ item : 
![image](https://user-images.githubusercontent.com/2470104/159227532-466cd5f2-f063-4a08-b73e-358f7590aa68.png)

Then, If I want to use the input to search a FAQ item, the list is empty : 
![image](https://user-images.githubusercontent.com/2470104/159227819-29772226-5e88-4f80-8059-e7306d2a213b.png)

**Comment by btry** [comment]
Hi

The error you fould is very likely the cause of your problem, but the lines are incomplete (did you copy / pasted a terminal running nano ?)

Please show the complete error as I need to read the SQL query.

**Comment by paidge** [comment]
Yes. Sorry.

```[2022-03-07 16:43:16] glpisqllog.ERROR: DBmysql::query() in /var/www/apps/glpi/inc/dbmysql.class.php line 335
  *** MySQL query error:
  SQL: SELECT `glpi_plugin_formcreator_forms`.`id`, `glpi_plugin_formcreator_forms`.`name`, `glpi_plugin_formcreator_forms`.`icon`, `glpi_plugin_formcreator_forms`.`icon_color`, `glpi_plugin_formcreator_forms`.`background_color`, `glpi_plugin_formcreator_forms`.`description`, `glpi_plugin_formcreator_forms`.`usage_count`, `glpi_plugin_formcreator_forms`.`is_default` FROM `glpi_plugin_formcreator_forms` LEFT JOIN `glpi_plugin_formcreator_categories` ON (`glpi_plugin_formcreator_categories`.`id` = `glpi_plugin_formcreator_forms`.`plugin_formcreator_categories_id`) LEFT JOIN `glpi_plugin_formcreator_sections` ON (`glpi_plugin_formcreator_sections`.`plugin_formcreator_forms_id` = `glpi_plugin_formcreator_forms`.`id`) LEFT JOIN `glpi_plugin_formcreator_questions` ON (`glpi_plugin_formcreator_questions`.`plugin_formcreator_sections_id` = `glpi_plugin_formcreator_sections`.`id`) LEFT JOIN `glpi_plugin_formcreator_forms_profiles` ON (`glpi_plugin_formcreator_forms_profiles`.`plugin_formcreator_forms_id` = `glpi_plugin_formcreator_forms`.`id`) LEFT JOIN `glpi_plugin_formcreator_forms_languages` ON (`glpi_plugin_formcreator_forms`.`id` = `glpi_plugin_formcreator_forms_languages`.`plugin_formcreator_forms_id`) WHERE (`glpi_plugin_formcreator_forms`.`is_active` = '1' AND `glpi_plugin_formcreator_forms`.`is_deleted` = '0' AND (`glpi_plugin_formcreator_forms`.`language` IN ('fr_FR', '0', '', NULL) OR `glpi_plugin_formcreator_forms_languages`.`name` = 'fr_FR') AND ((MATCH(glpi_plugin_formcreator_forms.`name`, glpi_plugin_formcreator_forms.`description`)
                  AGAINST('comment* on* s\\\'en* sert* ?* ;)*' IN BOOLEAN MODE) OR MATCH(glpi_plugin_formcreator_questions.`name`, glpi_plugin_formcreator_questions.`description`)
                  AGAINST('comment* on* s\\\'en* sert* ?* ;)*' IN BOOLEAN MODE))) AND ((`access_rights` != '2' OR (`glpi_plugin_formcreator_forms_profiles`.`profiles_id` = '1')))) GROUP BY `glpi_plugin_formcreator_forms`.`id`, `glpi_plugin_formcreator_forms`.`name`, `glpi_plugin_formcreator_forms`.`description`, `glpi_plugin_formcreator_forms`.`usage_count`, `glpi_plugin_formcreator_forms`.`is_default` ORDER BY `glpi_plugin_formcreator_forms`.`name` ASC
  Error: syntax error, unexpected ')', expecting $end
  Backtrace :
  inc/dbmysqliterator.class.php:95                   
  inc/dbmysql.class.php:902                          DBmysqlIterator->execute()
  plugins/formcreator/inc/form.class.php:983         DBmysql->request()
  plugins/formcreator/ajax/homepage_wizard.php:62    PluginFormcreatorForm->showFormList()
  plugins/formcreator/ajax/homepage_wizard.php:48    plugin_formcreator_showWizardForms()
  {"user":"2310@hd-dev"} ```

**Comment by btry** [comment]
I found the problem.

The plugin adds an asterisk after each word to allow searching for incomplete words in the input. However it is not smart enough to handle smileys.

**Comment by paidge** [comment]
I'm not sure this error is linked to my problem because [the precedent test](https://github.com/pluginsGLPI/formcreator/issues/2634#issuecomment-1073636221) did not generate an error in sql-errors.log

**Comment by btry** [comment]
Hi


First, here is a fix for the "not found" error message. It is used in all search types, then only the label was inexact.


I think I may find what's happening by getting the SQL queries of the 2 cases described in this comment https://github.com/pluginsGLPI/formcreator/issues/2634#issuecomment-1073636221

**Comment by btry** [comment]
Please apply the followign patch to log SQL queries in the file glpi/files/_log/sql-errors.log, then try the 3 tests you did and show the logged queries 

```diff
diff --git a/inc/knowbase.class.php b/inc/knowbase.class.php
index 3295866f..39a5d212 100644
--- a/inc/knowbase.class.php
+++ b/inc/knowbase.class.php
@@ -210,6 +210,7 @@ class PluginFormcreatorKnowbase {
 
       $formList = [];
       $result_faqs = $DB->request($query_faqs);
+      Toolbox::logSqlError($result_faqs->getSQL());
       foreach ($result_faqs as $faq) {
          $formList[] = [
             'id'           => $faq['id'],
```

**Comment by paidge** [comment]
I am sorry about that but, after testing again, it seems the problem does not occur anymore. I am not able to say why. The search works well for KB.

**Comment by btry** [comment]
I had recently feedback of malfunction of the KB part of he service catalog. In the case I observed, this was caused by an inconsistency on the KB categories in GLPI. The column completename contained NULL instead of a concatenation of the names of the categories from the root to the item.

Maybe this was what the reason.

If the issue does not occurs anymore, and having no valuable track to find how this happened, I think we can close.
